vllm/entrypoints/chat_utils.py

```diff
@@ -68,16 +68,21 @@ class CustomChatCompletionMessageParam(TypedDict, total=False):
     Provides the model information to differentiate between participants of the
     same role.
     """
+    tool_call_id: Optional[str]
+    tool_calls: Optional[List]
 
 
 ChatCompletionMessageParam = Union[OpenAIChatCompletionMessageParam,
                                    CustomChatCompletionMessageParam]
 
 
 # TODO: Make fields ReadOnly once mypy supports it
-class ConversationMessage(TypedDict):
+class ConversationMessage(TypedDict, total=False):
     role: str
-    content: str
+    content: Optional[str]
+    tool_call_id: Optional[str]
+    name: Optional[str]
+    tool_calls: Optional[List]
```

Comment:
You can use `typing_extensions.NotRequired` for individual fields instead of setting `total=False` which applies to all fields. This should avoid the need to `cast` the `role` later in the code.

Reply:
I tried this, but it didn't fix it. It actually caused a cascade of other type issues for some reason. I'm not sure why, but it seems to be inferring the type on the right-hand-side incredibly broadly, likely related to the multi-part content structure used for vision models, but the left-hand side is much narrower. 

This is the type inferred for `tool_call_id` on line 233: `tool_call_id: str | Iterable[ChatCompletionContentPartTextParam] | Literal["system"] | None`

Same for `tool_calls` on 234: `tool_calls: str | Iterable[ChatCompletionContentPartTextParam] | Literal["system"] | None` and `name` on 236: `name: str | Iterable[ChatCompletionContentPartTextParam] | Literal["system"]`

Obviously we know the type is narrower at runtime given the context, but the type isn't being narrowed properly based on the conditional I don't think. I know typescript does that (that's my wheelhouse), but I'm not quite as familiar with pydantic.

Reply:
Hmm... perhaps we need to help the type checker a bit by using a discriminated union.

Reply:
So the issue here is that the type in question is the `ChatCompletionMessageParam` on like 227. 
This type is  defined as:

```python
ChatCompletionMessageParam = Union[
    OpenAIChatCompletionMessageParam,
    CustomChatCompletionMessageParam
]
```

`CustomChatCompletionMessageParam` is defined in a straightforward way: 

```python
class CustomChatCompletionMessageParam(TypedDict, total=False):
    """Enables custom roles in the Chat Completion API."""
    role: Required[str]
    """The role of the message's author."""

    content: Union[str, List[ChatCompletionContentPartParam]]
    """The contents of the message."""

    name: str
    """An optional name for the participant.

    Provides the model information to differentiate between participants of the
    same role.
    """
    tool_call_id: Optional[str]
    tool_calls: Optional[List]
 ```

The issue that we get is how `OpenAIChatCompletionMessageParam` is defined -- it's renamed from this:
```python
from openai.types.chat import ChatCompletionMessageParam as OpenAIChatCompletionMessageParam
```

Now, `ChatCompletionMessageParam` in this case is defined in the OpenAI SDK, which is where the issue comes from (`openai.types.chat.chat_completion_message_param.py`), and which we can't easily discriminate:

```python
ChatCompletionMessageParam: TypeAlias = Union[
    ChatCompletionSystemMessageParam,
    ChatCompletionUserMessageParam,
    ChatCompletionAssistantMessageParam,
    ChatCompletionToolMessageParam,
    ChatCompletionFunctionMessageParam,
]
```

There might be a good way to add discrimination on top of this, but it's beyond me as I'm not super familiar with pydantic and Claude isn't very helpful here. Open to feedback though if you feel like there's a good way to solve this!

Reply:
Let me clone your fork and try to fix it.

Reply:
Looking into it further, it seems that you are confusing `FunctionCall.name` with `ChatCompletionAssistantMessageParam.name`.

Reply:
I have updated the parsing logic, but don't have permissions to push to your fork. Here's the diffs which you can apply locally:

```diff
diff --git a/vllm/entrypoints/chat_utils.py b/vllm/entrypoints/chat_utils.py
index 8e4c2a08..7845d3ff 100644
--- a/vllm/entrypoints/chat_utils.py
+++ b/vllm/entrypoints/chat_utils.py
@@ -1,21 +1,25 @@
 import codecs
 from dataclasses import dataclass
-from functools import lru_cache
+from functools import lru_cache, partial
 from pathlib import Path
 from typing import (Any, Awaitable, Iterable, List, Literal, Optional, Tuple,
                     Union, cast)
 
 # yapf conflicts with isort for this block
 # yapf: disable
-from openai.types.chat import ChatCompletionContentPartImageParam
+from openai.types.chat import (ChatCompletionAssistantMessageParam,
+                               ChatCompletionContentPartImageParam,
+                               ChatCompletionMessageToolCallParam,
+                               ChatCompletionToolMessageParam)
 from openai.types.chat import (
     ChatCompletionContentPartParam as OpenAIChatCompletionContentPartParam)
-from openai.types.chat import ChatCompletionContentPartTextParam
+from openai.types.chat import (ChatCompletionContentPartRefusalParam,
+                               ChatCompletionContentPartTextParam)
 from openai.types.chat import (
     ChatCompletionMessageParam as OpenAIChatCompletionMessageParam)
 # yapf: enable
 # pydantic needs the TypedDict from typing_extensions
-from pydantic import ConfigDict, TypeAdapter
+from pydantic import ConfigDict
 from typing_extensions import Required, TypeAlias, TypedDict
 
 from vllm.config import ModelConfig
@@ -51,7 +55,7 @@ class CustomChatCompletionContentPartParam(TypedDict, total=False):
 
 ChatCompletionContentPartParam: TypeAlias = Union[
     OpenAIChatCompletionContentPartParam, ChatCompletionContentPartAudioParam,
-    CustomChatCompletionContentPartParam, ]
+    ChatCompletionContentPartRefusalParam, CustomChatCompletionContentPartParam]
 
 
 class CustomChatCompletionMessageParam(TypedDict, total=False):
@@ -68,8 +72,12 @@ class CustomChatCompletionMessageParam(TypedDict, total=False):
     Provides the model information to differentiate between participants of the
     same role.
     """
+
     tool_call_id: Optional[str]
-    tool_calls: Optional[List]
+    """Tool call that this message is responding to."""
+
+    tool_calls: Optional[Iterable[ChatCompletionMessageToolCallParam]]
+    """The tool calls generated by the model, such as function calls."""
 
 
 ChatCompletionMessageParam = Union[OpenAIChatCompletionMessageParam,
@@ -78,11 +86,20 @@ ChatCompletionMessageParam = Union[OpenAIChatCompletionMessageParam,
 
 # TODO: Make fields ReadOnly once mypy supports it
 class ConversationMessage(TypedDict, total=False):
-    role: str
+    role: Required[str]
+    """The role of the message's author."""
+
     content: Optional[str]
+    """The contents of the message."""
+
     tool_call_id: Optional[str]
+    """Tool call that this message is responding to."""
+
     name: Optional[str]
-    tool_calls: Optional[List]
+    """The name of the function to call."""
+
+    tool_calls: Optional[Iterable[ChatCompletionMessageToolCallParam]]
+    """The tool calls generated by the model, such as function calls."""
 
 
 @dataclass(frozen=True)
@@ -155,9 +172,11 @@ def _get_full_multimodal_text_prompt(placeholder_token_str: str,
     return f"{placeholder_token_str}\n{text_prompt}"
 
 
-_TextParser = TypeAdapter(ChatCompletionContentPartTextParam)
-_ImageParser = TypeAdapter(ChatCompletionContentPartImageParam)
-_AudioParser = TypeAdapter(ChatCompletionContentPartAudioParam)
+# No need to validate using Pydantic again
+_TextParser = partial(cast, ChatCompletionContentPartTextParam)
+_ImageParser = partial(cast, ChatCompletionContentPartImageParam)
+_AudioParser = partial(cast, ChatCompletionContentPartAudioParam)
+_RefusalParser = partial(cast, ChatCompletionContentPartRefusalParam)
 
 
 def _parse_chat_message_content_parts(
@@ -173,7 +192,7 @@ def _parse_chat_message_content_parts(
     for part in parts:
         part_type = part["type"]
         if part_type == "text":
-            text = _TextParser.validate_python(part)["text"]
+            text = _TextParser(part)["text"]
             texts.append(text)
         elif part_type == "image_url":
             modality = "image"
@@ -181,7 +200,7 @@ def _parse_chat_message_content_parts(
                 raise NotImplementedError(
                     "Multiple multimodal inputs is currently not supported.")
 
-            image_url = _ImageParser.validate_python(part)["image_url"]
+            image_url = _ImageParser(part)["image_url"]
 
             if image_url.get("detail", "auto") != "auto":
                 logger.warning(
@@ -196,9 +215,12 @@ def _parse_chat_message_content_parts(
                 raise NotImplementedError(
                     "Multiple multimodal inputs is currently not supported.")
 
-            audio_url = _AudioParser.validate_python(part)["audio_url"]
+            audio_url = _AudioParser(part)["audio_url"]
             audio_future = async_get_and_parse_audio(audio_url["url"])
             mm_futures.append(audio_future)
+        elif part_type == "refusal":
+            text = _RefusalParser(part)["refusal"]
+            texts.append(text)
         else:
             raise NotImplementedError(f"Unknown part type: {part_type}")
 
@@ -223,53 +245,47 @@ def _parse_chat_message_content_parts(
     return ChatMessageParseResult(messages=messages, mm_futures=mm_futures)
 
 
+# No need to validate using Pydantic again
+_AssistantParser = partial(cast, ChatCompletionAssistantMessageParam)
+_ToolParser = partial(cast, ChatCompletionToolMessageParam)
+
+
 def _parse_chat_message_content(
     message: ChatCompletionMessageParam,
     model_config: ModelConfig,
     tokenizer: AnyTokenizer,
 ) -> ChatMessageParseResult:
-    role = cast(str, message["role"])  # can be iterable in some cases, so cast
+    role = message["role"]
     content = message.get("content")
-    tool_call_id = message.get("tool_call_id")
-    tool_calls = message.get("tool_calls", None)
-    # no longer used by OpenAI, but some models still use it for tool calls.
-    name = message.get("name", "")
-
-    # empty case
-    if content is None and tool_calls is None:
-        return ChatMessageParseResult(messages=[], mm_futures=[])
-
-    # special case - assistant message where tool calls are provided.
-    if role == "assistant" and tool_calls is not None:
-        messages = [
-            ConversationMessage(role=role,
-                                content=cast(Optional[str], content),
-                                tool_calls=list(tool_calls))
-        ]
-        return ChatMessageParseResult(messages=messages, mm_futures=[])
-
-    # special case - tool call result message
-    elif role == "tool":
-        messages = [
-            ConversationMessage(role=role,
-                                name=name,
-                                content=cast(Union[str, None], content),
-                                tool_call_id=cast(Union[str, None],
-                                                  tool_call_id))
-        ]
-        return ChatMessageParseResult(messages=messages, mm_futures=[])
-
-    # other cases - normal assistant response, user message or system message
+
+    if content is None:
+        result = ChatMessageParseResult(messages=[], mm_futures=[])
     elif isinstance(content, str):
         messages = [ConversationMessage(role=role, content=content)]
-        return ChatMessageParseResult(messages=messages, mm_futures=[])
-
-    return _parse_chat_message_content_parts(
-        role,
-        content,  # type: ignore
-        model_config,
-        tokenizer,
-    )
+        result = ChatMessageParseResult(messages=messages, mm_futures=[])
+    else:
+        result = _parse_chat_message_content_parts(
+            role,
+            content,  # type: ignore
+            model_config,
+            tokenizer,
+        )
+
+    for result_msg in result.messages:
+        if role == "assistant":
+            parsed_msg = _AssistantParser(message)
+
+            if (function_call := parsed_msg.get("function_call")) is not None:
+                result_msg["name"] = function_call["name"]
+            if "tool_calls" in parsed_msg:
+                result_msg["tool_calls"] = parsed_msg["tool_calls"]
+        elif role == "tool":
+            parsed_msg = _ToolParser(message)
+
+            if "tool_call_id" in parsed_msg:
+                result_msg["tool_call_id"] = parsed_msg["tool_call_id"]
+
+    return result
 
 
 def parse_chat_messages(
```

Reply:
Thanks! I will apply this patch this evening and re-run tests

Reply:
> Looking into it further, it seems that you are confusing `FunctionCall.name` (in assistant message btw, not tool message) with `CustomChatCompletionMessageParam.name` (which has nothing to do with Tools API, I think).

Yes and no - the `CustomChatCompletionMessageParam.name` is actually used by Hermes in tool response dicts:

```jinja
 {%- elif message.role == "tool" %}
        {%- if not message.name is defined %}
            {{- raise_exception("Tool response dicts require a 'name' key indicating the name of the called function!") }}
        {%- endif %}
        {%- if loop.previtem and loop.previtem.role != "tool" %}
            {{- '<|im_start|>tool\n' }}
        {%- endif %}
        {{- '<tool_response>\n' }}
        {{- message.content }}
        {%- if not loop.last %}
            {{- '\n</tool_response>\n' }}
        {%- else %}
            {{- '\n</tool_response>' }}
        {%- endif %}
        {%- if not loop.last and loop.nextitem.role != "tool" %}
            {{- '<|im_end|>' }}
        {%- elif loop.last %}
            {{- '<|im_end|>' }}
        {%- endif %}
    {%- endif %}
```
Basically, it relies on you setting a `name` property on `tool`-role messages to the name of the tool that was called, so that it can be used in the chat template. 

Your changes to the code above result in two tests failing for Hermes only, not Mistral:
```
FAILED tests/tool_use/test_parallel_tool_calls.py::test_parallel_tool_calls_with_results[hermes] - openai.BadRequestError: Error code: 400 - {'object': 'error', 'message': "Tool response dicts require a 'name' key indicating the name of the called function!", 'type': 'BadRequestError', 'param': N...
FAILED tests/tool_use/test_tool_calls.py::test_tool_call_with_results[hermes] - openai.BadRequestError: Error code: 400 - {'object': 'error', 'message': "Tool response dicts require a 'name' key indicating the name of the called function!", 'type': 'BadRequestError', 'param': N...
```
(Note that the error message matches the exception raised in the chat template). 

Since the `name` field is not required in `tool`-role messages any more [per the OpenAI Spec](https://platform.openai.com/docs/api-reference/chat/create), there are a couple ways to approach solving this:
1. _allow_ the `name` field to be set on `tool`-role messages. This may be required by other models in the future as well since they often rely on the name of the tool that was called instead of the ID of the tool call (like Mistral), so it may not be a bad idea. 
2. in vLLM's source code, look up the name of the function that was called based a the previous `assistant` message with `tool_calls` in it with a tool call ID that matches the `tool_call_id` specified in the `tool`-role message, and set that on the tool dict before passing it into the chat template
3. do the same thing as above, but in the hermes chat template rather than in the vLLM source code. This would add a _lot_ of complexity, but is probably better than option # 2

My suggestion:
implement allowing the `name` property to be set on `tool`-role messages for ease of integration per # 1. 
Alternatively, # 3 is the next-best option. 

cc @interstellarninja

Reply:
Ahh I understand better now. In the chat template, `tool_call.name` is accessed for `message.role == "assistant"` but not `message.role == "tool"`. Is that intended?

Reply:
This code is still breaking tests @DarkLight1337 it is resulting in assistant messages with `tool_calls` not being passed into the template properly I think -- digging into it further.

Reply:
For now, we can edit the for loop so that `message.name` is unconditionally added to the result if it exists.

```diff
     for result_msg in result.messages:
         if role == "assistant":
             parsed_msg = _AssistantParser(message)
 
-            if (function_call := parsed_msg.get("function_call")) is not None:
-                result_msg["name"] = function_call["name"]
             if "tool_calls" in parsed_msg:
                 result_msg["tool_calls"] = parsed_msg["tool_calls"]
         elif role == "tool":
             parsed_msg = _ToolParser(message)
 
             if "tool_call_id" in parsed_msg:
                 result_msg["tool_call_id"] = parsed_msg["tool_call_id"]

+        if (name := message.get("name")) is not None:
+            result_msg["name"] = name
```

Reply:
ok - I am working on fixing the other tests that the diff broke, it seems like it's not parsing tool call messages or tool call result message properly.

Reply:
I think I fixed it - basically I had to check  EITHER if `content` is a string, OR if `tool_calls` is defined. But, to narrow it properly for the type checker and avoid the multipart issue, I had to add into the condition a check that if `tool_calls` is defined and `content` is a string or `None`.

Reply:
At this point, perhaps it would be simpler to normalize `content` into a list of messages so that we can call `_parse_chat_message_content_parts` and subsequent processing unconditionally...

Reply:
> Alternatively, the lines with the exception about `message.name` not being defined in the chat template can safely be omitted from the chat template - because, the `message.name` isn't actually used contrary to the error message.
> 
> @interstellarninja can you confirm?
> Portion that's safe to remove:
> 
> ```
>         {%- if not message.name is defined %}
>             {{- raise_exception("Tool response dicts require a 'name' key indicating the name of the called function!") }}
>         {%- endif %}
> ```

yes you can remove the requirement for message.name to be provided since `name` has been replaced in favor of `tool_call_id` in the OpenAI API Spec. 

however ideally we would prefer function name to be passed as part of `tool` message. since we are not able to pass name through the api call, we could perhaps infer from the tools called using the id in the jinja as you suggested.

one hack developers can use is to pass function name as part of content string which is not easy to standardize.

Reply:
> At this point, perhaps it would be simpler to normalize `content` into a list of messages so that we can call `_parse_chat_message_content_parts` and subsequent processing unconditionally...

I'm not quite sure what you mean @DarkLight1337 - can you explain a little more? And, is this something you want to see as part of this PR?

Reply:
Since this PR is modifying the parsing logic, best to avoid unnecessary complexity. I'm thinking of something like this:

```py
    if content is None:
        content = []
    if isinstance(content, str):
        content = [ChatCompletionContentPartTextParam(type="text",
                                                      text=content)]

    result = _parse_chat_message_content_parts(
        role,
        content,  # type: ignore
        model_config,
        tokenizer,
    )

    # The rest is the same
```

By the way, this shouldn't be needed anymore since `name` should be based on `message.name`.
```diff,py
 if (function_call := parsed_msg.get("function_call")) is not None:
               result_msg["name"] = function_call["name"]
```

Reply:
Ok I think I see - and then `parse_chat_message_content_parts` would handle any other logic that has to deal with the `content`?

Reply:
Yep!

Reply:
Done.